File: 0042. Trapping Rain Water/Solution0042.py
from typing import List


class Solution0042:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height)-1
        leftmax, rightmax = height[left], height[right]
        sum = 0
        while left <= right:
            if leftmax <= rightmax:
                sum += max(leftmax - height[left], 0)
                leftmax = max(leftmax, height[left])
                left += 1
            else:
                sum += max(rightmax - height[right], 0)
                rightmax = max(rightmax, height[right])
                right -= 1
        return sum

File: 0042. Trapping Rain Water/Solution0042.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0, right = height.size()-1, lMax = height[left], rMax = height[right];
        int water = 0;
        while (left < right) {
            lMax = max(lMax, height[left]);
            rMax = max(rMax, height[right]);
            if (height[left] < height[right]) {
                water += lMax - height[left];
                left++;
            } else {
                water += rMax - height[right];
                right--;
            }
        }
        return water;        
    }
};